### ISSUE TITLE
Inconsistent window size on HiDPI screen + incorrect font scaling on dpi switch

### ISSUE THREAD

**Issue by lompabo** [jwilm/alacritty#1382]
Which operating system does the issue occur on?

OsX, on a retina display with an attached (non-retina) secondary monitor

I am on the master branch, commit: f55252ec8537e68d0c98a1fac35728a2b596d328

Using the (default) configuration options:

```
window:
  dimensions:
    columns: 80
    lines: 24

scale_with_dpi: true
```

The default alacrity window is much smaller than it should be. Here's a screenshot comparing an 80x24 Terminal.app with an 80x24 alacrity window

<img width="959" alt="screen shot 2018-06-18 at 10 19 14" src="https://user-images.githubusercontent.com/5887142/41525384-175496ba-72e1-11e8-9f2f-693a24061477.png">

The screenshot has been obtained running alacrity with the -v flag, so that the window on the left (Terminal.app) contains also the log.

The device pixel ratio seems to be detected correctly (2x on retina displays). I cannot really make full sense of the first "width: ..., height: ..." line (is the size of glyph cache?), but the third (562x338) seems to be just the second (1124x676), scaled down according to the pixel ratio.

My first thought was that the window size was computed prior to the dpi scaling. If that was the case, I would expect that doubling the windows size (in terms of lines and columns) should get a correctly sized window. In practice, doubling up the size...

```
window:
  dimensions:
    columns: 160
    lines: 48
```

...gets very close to Terminal.app 80x24. There is still a slight difference (no need for a screenshot here), but I figure that could be due to minor differences in cell size between the two emulators.

Now, moving the alacrity window on a secondary (non-retina) monitor keeps its size (correctly), but results in double-sized fonts. Here is a screenshot of the two windows, once they have been moved to the secondary monitor.

![screen shot 2018-06-18 at 10 42 51](https://user-images.githubusercontent.com/5887142/41526395-67e03a1e-72e4-11e8-960b-3dc38e6ed006.png)

The log (on the left-side window) shows that moving alacrity to the secondary monitor triggered a size update.
```
width: 281, height: 169
```

Just for the record, repeating the process backwards (run on the secondary -- non retina -- monitor, move to the primary -- retina -- monitor) yields reverse results. Here's what happens when first running alacrity (the window has the correct size):

![screen shot 2018-06-18 at 10 49 09](https://user-images.githubusercontent.com/5887142/41526690-482419ec-72e5-11e8-952e-de655afd908c.png)

And here is what happens when moving both windows on the retina display (window size is kept, half size fonts):

<img width="1184" alt="screen shot 2018-06-18 at 10 51 49" src="https://user-images.githubusercontent.com/5887142/41526827-a10b544e-72e5-11e8-866d-e02940802729.png">

Again, the move triggered a size update.

In summary:

- When the window is created, the font size si always correct, but the window size is halved on retina displays
- When the switching to a different monitor, the window size is correctly handled, but the font size is either doubled or halved.

Even  if it looks like to issues, I suspect they are both related to the code for handling the same event. Unfortunately I am not familiar with rust, so I am not sure I would be able to provide more information by looking at the code.

Michele

BTW: guys & gals, many thanks for all your great work!

**Comment by monouser7dig** [comment]
I have the same behaviour on 13 inch retina MacBook without external monitor.

**Comment by chrisduerr** [comment]
Okay, let's tackle this one thing after the other.

Currently there DPI does not readjust when the monitor is switched, this has been blocked on one of our dependencies but will likely soon be available on all platforms. There's already an open PR for MacOS (#1346).

Regarding the window width and height, this is calculated as `cell_width * cols + 2 * padding`, so I'm not sure why this is not correct. I'm unfortunately not able to test anything on macos.

**Comment by chrisduerr** [comment]
Thanks for testing the PR! It should hopefully fix all remaining DPI issues.

**Comment by laanwj** [comment]
just encountered the same problem, have to set
```
window:
  dimensions:
    columns: 121
    lines: 37
```
to get 80×24 terminal
this is only necessary on a laptop on which the DPI is non-standard

will test the suggested solutions
